### DISCUSSION TITLE
Razorpay Integration | India

### DISCUSSION THREAD

**Discussion by decorsstudio** [swellstores/community#255] [Help] [answered]
Hello Members,
I am trying to add payment gateway as [Razorpay ](https://razorpay.com/)for my store, but seems there is no such type of integration for that.
I am trying to build the app which will be available in India only for now, and our requirement is to use [Razorpay ](https://razorpay.com/)payment gateway only. 

Please help if there is an option to setup it, other wise I have to choose another platform. 
For our project Swell Is good fit on the required features, but only issue with the payment platform. please help if it possible.

**Comment by logeshswell** [comment, marked as answer]
Swell supports numerous payment gateways, wallets, and alternative payment methods such as Stripe, Braintree, PayPal and more. However, in certain cases, you may want to integrate Swell with a custom payment gateway or method that is not already supported. 

For example, you may want to integrate [Razorpay](https://razorpay.com/) payment gateway to accept payments. The following guide helps to configure and manage a custom payment gateway (or method), create payments, and handle refunds.

**Reply by logeshswell** [reply]
The first step is to create a custom payment method following the below steps:
1. Login to the Swell dashboard and navigate to **Settings** > **Payments**. 
2. Under **Manual payments**, click **Add manual payment method** button.
3. Select 'Custom' in the payment method dropdown.
4. Input the payment method name and ID fields as shown in the screenshot and Click **Save** button.
  a. This is the ID you will use in the frontend or backend APIs to update the cart billing method and create payments.

<img width="670" alt="Screenshot 2023-05-21 at 10 21 06 PM" src="https://github.com/swellstores/community/assets/100766131/083dc37b-9034-4410-b132-34c94e5b8ef5">

### Note:

* This custom payment method can be used in both Hosted and Custom checkouts.
* In Hosted Checkout, they’ll appear as payment methods under the Payment method section as follows:
<img width="608" alt="Screenshot 2023-05-21 at 10 40 07 PM" src="https://github.com/swellstores/community/assets/100766131/1d839549-5e64-40f6-bf98-ad44942ac926">

* Customer can choose this custom payment method and complete the checkout.
* If you require additional payment info from the customer, you can redirect the customer to a custom page and capture those details.
  * To do so, Navigate to **Storefronts** > **Checkout** and enable the Custom return link toggle to input the redirect URL.

**Reply by logeshswell** [reply]
In Custom checkout, you can update the cart’s billing method using the payment method ID that we created earlier and the additional payment info inside `metadata` field as follows:
```ts
await swell.cart.update({
  billing: {
    method: 'razorpay'
  },
  metadata: {
    razorpay: {
      order_id: 'rzp_13r432qws'
    }
  }
});
```

**Reply by logeshswell** [reply]
The second step is to create payments in Swell.

Submitting the cart with `swell.cart.submitOrder()` will create an order with the billing method and metadata that was updated in the cart. By default, orders created with a custom payment method won't have any payments associated with them and will have an `unpaid` order status in the Swell dashboard.

Depending on the intended payment flow, you can use a webhook firing on the `order.created` event, or await for the response of `submitOrder()` call in order to create payments.

To create payments, you’ll have to use the [Backend API](https://developers.swell.is/backend-api/payments/create-a-payment) by posting to the `/payments` endpoint as follows:
```ts
const swell = require('swell-node').init('STORE_ID', 'SECRET_KEY');

await swell.post('/payments', {
  amount: 10,
  method: 'razorpay',
  order_id: '6425a559c5aaf00012670aa9',
  account_id: '60a45168c19da965a5b399de',
  captured: true,
  authorized: true,
  razorpay_order_id: 'rzp_13r432qws' // custom-field that can be referenced from order.metadata
});
```

**Reply by logeshswell** [reply]
If you want to build a typical Auth-n-capture flow, follow the below steps:
1. Create a payment with `authorized: true` and `captured: false` to ensure that you’re not capturing the payment.
  a. This would create a payment with the status 'Authorized' instead of 'Paid' in the Swell dashboard.

<img width="875" alt="Screenshot 2023-04-05 at 10 21 41 AM" src="https://github.com/swellstores/community/assets/100766131/9446b49b-08a3-4c40-925a-0e2a90b7bf6a">

2. When the charge is captured, update the payment with `captured: true` as follows:
```ts
const swell = require('swell-node').init('STORE_ID', 'SECRET_KEY');

await swell.put('/payments/<payment_id>', {
  captured: true
});
```
3. Now, you should see the order and payment statuses as `Paid`.

**Reply by logeshswell** [reply]
The third and last step is to setup a refund flow.

When you perform a refund action from the Swell dashboard, you will be triggered `order.refunded` webhook which you suppose to be configured following the [webhooks guide](https://developers.swell.is/guides/core-concepts/webhook-basics#creating-a-webhook).
``` json
{
  "user_id": "6059f83ea7a1d765d211b872",
  "type": "order.refunded",
  "data": {
    "payment_id": "642c671c57df50001226dfa9",
    "payment_refund_id": "642c672603fc0c0012735871",
    "id": "6426e581239bc700125dabac"
  },
  "model": "orders",
  "date_created": "2023-04-04T18:06:31.311Z",
  "date_updated": "2023-04-04T18:06:31.413Z",
  "webhooks_pending": 0,
  "id": "642c672703fc0c0012735878"
}
```

Querying the `/payments` endpoint with the received `payment_id` would retrieve all the necessary data that was passed when the payment was created and it can be used to process the refund action in the payment gateway.

**Reply by logeshswell** [reply]
Swell’s architecture is built with flexibility in mind, especially when it comes to payments. As we’ve seen in the examples above, it’s easy to use Swell’s Frontend and Backend APIs in order to integrate a custom payment gateway (or method).

So, please feel free to comment in this thread if you have any doubts or clarifications.

**Reply by logeshswell** [reply]
cc: @decorsstudio

**Reply by decorsstudio** [reply]
Thanks for quick reply @logeshswell
